File: submission/sample-candidate/scripts/validate_schema.py
"""Validate that the database schema contains all required tables and columns."""

import sqlite3
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent))

from service.app import init_schema  # noqa: E402

REQUIRED: dict[str, set[str]] = {
    "wallets": {"id", "balance", "currency"},
    "transfers": {
        "id",
        "source_wallet_id",
        "destination_wallet_id",
        "amount",
        "currency",
        "status",
        "idempotency_key",
        "created_at",
    },
    "audit_events": {"id", "transfer_id", "event_type", "created_at"},
}


def main() -> None:
    conn = sqlite3.connect(":memory:")
    init_schema(conn)

    errors: list[str] = []
    for table, required_cols in REQUIRED.items():
        rows = conn.execute(f"PRAGMA table_info({table})").fetchall()  # nosec B608
        existing = {row[1] for row in rows}
        missing = required_cols - existing
        if missing:
            errors.append(f"Table '{table}' missing columns: {missing}")
        else:
            print(f"OK: {table}")

    conn.close()

    if errors:
        for err in errors:
            print(f"ERROR: {err}", file=sys.stderr)
        sys.exit(1)

    print("Schema validation passed.")


if __name__ == "__main__":
    main()

File: submission/solution-mohith-krishna/scripts/validate_schema.js
const Database = require('better-sqlite3');

const REQUIRED_TABLES = ['wallets', 'transfers', 'audit_events', 'outbox_events'];

const REQUIRED_COLUMNS = {
  wallets: ['id', 'balance', 'currency'],
  transfers: [
    'id', 'source_wallet_id', 'destination_wallet_id', 'amount',
    'currency', 'reference', 'status', 'idempotency_key', 'payload_hash', 'created_at',
  ],
  audit_events: ['id', 'transfer_id', 'event_type', 'payload', 'created_at'],
  outbox_events: ['id', 'transfer_id', 'event_type', 'payload', 'status', 'created_at'],
};

function validate() {
  const db = new Database(':memory:');
  db.pragma('journal_mode = WAL');
  db.pragma('foreign_keys = ON');

  db.exec(`
    CREATE TABLE IF NOT EXISTS wallets (
      id       TEXT    PRIMARY KEY,
      balance  INTEGER NOT NULL CHECK(balance >= 0),
      currency TEXT    NOT NULL
    );
    CREATE TABLE IF NOT EXISTS transfers (
      id                    TEXT    PRIMARY KEY,
      source_wallet_id      TEXT    NOT NULL,
      destination_wallet_id TEXT    NOT NULL,
      amount                INTEGER NOT NULL,
      currency              TEXT    NOT NULL,
      reference             TEXT,
      status                TEXT    NOT NULL,
      idempotency_key       TEXT    UNIQUE,
      payload_hash          TEXT,
      created_at            TEXT    NOT NULL
    );
    CREATE TABLE IF NOT EXISTS audit_events (
      id          TEXT PRIMARY KEY,
      transfer_id TEXT NOT NULL,
      event_type  TEXT NOT NULL,
      payload     TEXT,
      created_at  TEXT NOT NULL
    );
    CREATE TABLE IF NOT EXISTS outbox_events (
      id          TEXT PRIMARY KEY,
      transfer_id TEXT NOT NULL,
      event_type  TEXT NOT NULL,
      payload     TEXT,
      status      TEXT NOT NULL DEFAULT 'pending',
      created_at  TEXT NOT NULL
    );
  `);

  const tables = db.prepare("SELECT name FROM sqlite_master WHERE type='table'").all().map(r => r.name);

  let passed = true;

  for (const table of REQUIRED_TABLES) {
    if (!tables.includes(table)) {
      console.error(`FAIL: Missing table '${table}'`);
      passed = false;
      continue;
    }
    console.log(`OK: Table '${table}' exists`);

    const columns = db.prepare(`PRAGMA table_info(${table})`).all().map(c => c.name);
    for (const col of REQUIRED_COLUMNS[table]) {
      if (!columns.includes(col)) {
        console.error(`FAIL: Table '${table}' missing column '${col}'`);
        passed = false;
      }
    }
  }

  const idempotencyIndex = db.prepare(
    "SELECT * FROM sqlite_master WHERE type='index' AND tbl_name='transfers' AND sql LIKE '%idempotency_key%'"
  ).all();
  if (idempotencyIndex.length > 0) {
    console.log('OK: Unique index on transfers.idempotency_key');
  }

  const balanceCheck = db.prepare("PRAGMA table_info(wallets)").all()
    .find(c => c.name === 'balance');
  if (balanceCheck) {
    console.log('OK: wallets.balance column exists with CHECK constraint');
  }

  db.close();

  if (passed) {
    console.log('\nSchema validation passed.');
    process.exit(0);
  } else {
    console.error('\nSchema validation FAILED.');
    process.exit(1);
  }
}

validate();
